autoPyTorch/evaluation/train_evaluator.py

```diff
@@ -0,0 +1,340 @@
+                    Y_valid_preds = np.nanmean(Y_valid_preds, axis=0)
+            else:
+                Y_valid_preds = None
+
+            if self.X_test is not None:
+                Y_test_preds = np.array([Y_test_pred[i]
+                                         for i in range(self.num_folds)
+                                         if Y_test_pred[i] is not None])
+                # Average the predictions of several models
+                if len(Y_test_preds.shape) == 3:
+                    Y_test_preds = np.nanmean(Y_test_preds, axis=0)
+            else:
+                Y_test_preds = None
+
+            self.Y_optimization = Y_targets
+            self.Y_actual_train = Y_train_targets
+
+            self.model = self._get_model()
+
+            status = StatusType.SUCCESS
+            self.logger.debug("In train evaluator fit_predict_and_loss, loss:{}".format(opt_loss))
+            self.finish_up(
+                loss=opt_loss,
+                train_loss=train_loss,
+                opt_pred=Y_optimization_preds,
+                valid_pred=Y_valid_preds,
+                test_pred=Y_test_preds,
+                additional_run_info=additional_run_info,
+                file_output=self.num_folds == 1,
+                status=status,
+            )
+
+    def _fit_and_predict(self, model: BaseEstimator, fold: int, train_indices: Union[np.ndarray, List],
+                         test_indices: Union[np.ndarray, List],
+                         add_model_to_self: bool
+                         ) -> Tuple[np.ndarray, np.ndarray, Optional[np.ndarray], Optional[np.ndarray]]:
+
+        self.indices[fold] = ((train_indices, test_indices))
+
+        # Do only output the files in the case of iterative holdout,
+        # In case of iterative partial cv, no file output is needed
+        # because ensembles cannot be built
+        X = {'train_indices': train_indices,
+             'val_indices': test_indices,
+             'split_id': fold,
+             **self._init_params}  # fit dictionary
+        y = None
+        fit_and_suppress_warnings(self.logger, model, X, y)
+        self.logger.info("Model fitted, now predicting")
+        (
+            Y_train_pred,
+            Y_opt_pred,
+            Y_valid_pred,
+            Y_test_pred
+        ) = self._predict(
+            model,
+            train_indices=train_indices,
+            test_indices=test_indices,
+        )
+
+        if add_model_to_self:
+            self.model = model
+        else:
+            self.models[fold] = model
+
+        return Y_train_pred, Y_opt_pred, Y_valid_pred, Y_test_pred
+
+    def _predict(self, model: BaseEstimator,
+                 test_indices: Union[np.ndarray, List],
+                 train_indices: Union[np.ndarray, List]
+                 ) -> Tuple[np.ndarray, np.ndarray, Optional[np.ndarray], Optional[np.ndarray]]:
+
+        X_train = self.X_train
+        if isinstance(X_train, pd.DataFrame):
```

Comment:
I think we need to do something else here. we cannot just cast to numpy. This is something that the pipeline should handle. Casting to numpy will remove strings, that the pipeline might expect.

Reply:
the error is in accessing the data via indices if it is a dataframe we need to call iloc on it rather than how it is done right now. Also, I tried casting to numpy and it didn't change the data frame, it just made an np array with dtype=object

Reply:
Looks like `_predict_regression` just calls `pipeline.predict`. I would assume that that can handle whatever comes from the dataset class (`pd.DataFrame` or `np.ndarray`) and `X_train` here also comes from the dataset. So the issue is just in line 286/289. That we can handle with and if-else or sth like:

```
X_subsampler = lambda x: X_train[x] if isinstance(X_train, np.ndarray) else X_train.iloc[x]
train_pred = self.predict_function(X_subsampler(train_indices), ...)
```

Reply:
Thanks for the suggestions, looks good to me, I'll make it this way then